include/DataflowPass.h

```diff
@@ -106,7 +107,72 @@ class DataflowPass {
 
     here, annoFnName = "free" and invokerFnName = "free0"
     */
-    virtual void onAnnotation(PVAliasSet* input, std::string &annoFnName, std::string& invokerFnName,
+    virtual void onAnnotation(PVAliasSet* input, const std::string &annoFnName, const std::string& invokerFnName,
+                              AnnotationType annotationType) = 0;
+
+    // annofnName is the name of the function inside the parameters.
+    // this overload exists (where we do not specify the function that invoked
+    // the call) because there are cases where it is not appropriate to add
+    // the invoker function to the called methods set
+    /*
+    Example: see the following code
+
+    -----------
+    struct my_struct {
+        char* x MustCall("free");
+        char* y MustCall("free");
+    };
+
+    my_struct Calls("free", "x") free_and_obligate(my_struct S Calls("free", "x")) {
+      free(S.x);
+      return S;
+    }
+
+    int main() {
+      my_struct A;
+      A.x = (char*)malloc(15);
+
+      my_struct B = free_and_obligate(A);
+    }
+    -----------
+
+    This code finds that there is a ReturnAnnotation on the function
+    free_and_obligate and will add "free" to return.x's called methods. Here, B is
+    set to that return value so B.x will have called methods {"free"}. But if we
+    also say to add the invoker name to the called methods (free_and_obligate),
+    then B.x's called methods will be {"free", "free_and_obligate"}. This is not
+    correct, hence the need to be able to include or and not include
+    the invoker name.
+
+    The only cases where invoker name needs to be specified are with
+    ParameterAnnotations & unknwon function
+    calls for the called methods pass and for
+    (as of the existence of Calls and MustCall annotations)
+
+
+    Most of these cases look like this:
+
+    -----------
+    struct my_struct {
+        char* x MustCall("free");
+        char* y MustCall("free");
+    };
+
+    void does_free(char* s Calls("free")) {
+      free(s);
+    }
+
+    int main() {
+      char* a = (char*)malloc(15);
+      does_free(a);
+    }
+    -----------
+
+    Here, it is appropriate that the called methods
+    of a = {"free", "does_free"}
```

Comment:
Agreed, but I'd expect this to happen due to two different calls (one for the annotation, and one for the dataflow fact) that are then GLB'd.